gpytorch/models/deep_gps/predictive_deep_gp.py

```diff
@@ -0,0 +1,119 @@
+from .deep_gp import DeepGPLayer as AbstractDeepGPLayer
+
+
+class AbstractPredictiveDeepGPLayer(AbstractDeepGPLayer):
+    def __init__(self, variational_strategy, input_dims, output_dims, num_sample_sites=3, quad_grid=None):
+        super().__init__(variational_strategy, input_dims, output_dims)
+
+        self.num_sample_sites = num_sample_sites
+
+        # Pass in previous_layer.quad_grid if you want to share quad_grid across layers.
+        if quad_grid is not None:
+            self.quad_grid = quad_grid
+        else:
+            self.quad_grid = torch.nn.Parameter(torch.randn(num_sample_sites, input_dims))
+
+    def __call__(self, inputs, are_samples=False, expand_for_quadgrid=True, **kwargs):
+        if isinstance(inputs, MultitaskMultivariateNormal):
+            # inputs is definitely in the second layer, and mean is n x t
+            mus, sigmas = inputs.mean, inputs.variance.sqrt()
+
+            if expand_for_quadgrid:
+                xi_mus = mus.unsqueeze(-3)  # 1 x n x t
+                xi_sigmas = sigmas.unsqueeze(-3)  # 1 x n x t
+            else:
+                xi_mus = mus
+                xi_sigmas = sigmas
+
+            # unsqueeze sigmas to 1 x n x t, locations from [q] to Q^T x 1 x T.
+            # Broadcasted result will be Q^T x N x T
+            qg = self.quad_grid.unsqueeze(-2)
+            # qg = qg + torch.randn_like(qg) * 1e-2
+            xi_sigmas = xi_sigmas * qg
+
+            inputs = xi_mus + xi_sigmas  # q^t x n x t
+        if settings.debug.on():
+            if not torch.is_tensor(inputs):
+                raise ValueError(
+                    "`inputs` should either be a MultitaskMultivariateNormal or a Tensor, got "
+                    f"{inputs.__class__.__Name__}"
+                )
+
+            if inputs.size(-1) != self.input_dims:
+                raise RuntimeError(
+                    f"Input shape did not match self.input_dims. Got total feature dims [{inputs.size(-1)}],"
+                    f" expected [{self.input_dims}]"
+                )
+
+        # Repeat the input for all possible outputs
+        if self.output_dims is not None:
+            inputs = inputs.unsqueeze(-3)
+            inputs = inputs.expand(*inputs.shape[:-3], self.output_dims, *inputs.shape[-2:])
+        # Now run samples through the GP
+        output = ApproximateGP.__call__(self, inputs, **kwargs)
+
+        if self.num_sample_sites > 0:
+            if self.output_dims is not None and not isinstance(output, MultitaskMultivariateNormal):
+                mean = output.loc.transpose(-1, -2)
+                covar = BlockDiagLazyTensor(output.lazy_covariance_matrix, block_dim=-3)
+                output = MultitaskMultivariateNormal(mean, covar, interleaved=False)
+        else:
+            output = output.loc.transpose(-1, -2)  # this layer provides noiseless kernel interpolation
+
+        return output
+
+
+class DeepPredictiveGaussianLikelihood(GaussianLikelihood):
```

Comment:
Given that `DeepLikelihood` is deprecated (in favor of `DeepApproximateMLL`), I would get rid of this class.

Reply:
I think for DSPPs the likelihood makes more sense than it did in the standard deep GP case because it currently handles summing over the quadrature weights (and is no longer just a matter of dealing with the additional samples dimension).

What do you think?

Reply:
I guess we could make like a `DeepPredictiveLogLikelihood` or something similar

Reply:
I'd be more in favor of `DeepPredictiveLogLikelihood` - otherwise we'll have a DSPPGaussianLikelihood and a DSPPMultitaskGaussianLikelihood and maybe a DSPPBernoulliLikelihood and DSPPSoftmaxLikelihood if we end up doing classification.

Reply:
@gpleiss I looked at this a bit more to refresh my memory, and we need the quadrature weights to use the predictions of the DSPP. E.g., the `log_marginal` code here gives the marginal predictive log likelihood for the input test point.

We need the quadrature grid to live somewhere, and it shouldn't be on the loss object since that doesn't make much sense for users to save and load. Moreover, I'm not sure it really makes sense for them to live on any particular DSPP layer. 

What I think I'm going to do is make a `DSPP(GP):` class analogous to the `DeepGP` container in `deep_gp.py` and put them there.